x-pack/plugins/task_manager/server/queries/task_claiming.ts

```diff
@@ -0,0 +1,459 @@
+      // but we do not apply this limitation on tasks claimed by ID so that we can
+      // provide more detailed error messages when we fail to claim them
+      (doc) => doc.status === TaskStatus.Claiming
+    );
+
+    // count how many tasks we've claimed by ID and validate we have capacity for them to run
+    const remainingCapacityOfClaimByIdByType = mapValues(
+      // This means we take the tasks that were claimed by their ID and count them by their type
+      countBy(documentsClaimedById, (doc) => doc.taskType),
+      (count, type) => this.getCapacity(type) - count
+    );
+
+    const [documentsClaimedByIdWithinCapacity, documentsClaimedByIdOutOfCapacity] = partition(
+      documentsClaimedById,
+      (doc) => {
+        // if we've exceeded capacity, we reject this task
+        if (remainingCapacityOfClaimByIdByType[doc.taskType] < 0) {
+          // as we're rejecting this task we can inc the count so that we know
+          // to keep the next one returned by ID of the same type
+          remainingCapacityOfClaimByIdByType[doc.taskType]++;
+          return false;
+        }
+        return true;
+      }
+    );
+
+    const documentsRequestedButNotReturned = difference(
+      claimTasksById,
+      documentsReturnedById.map((doc) => doc.id)
+    );
+
+    this.emitEvents([
+      ...documentsClaimedByIdWithinCapacity.map((doc) => asTaskClaimEvent(doc.id, asOk(doc))),
+      ...documentsClaimedByIdOutOfCapacity.map((doc) =>
+        asTaskClaimEvent(
+          doc.id,
+          asErr({
+            task: some(doc),
+            errorType: TaskClaimErrorType.CLAIMED_BY_ID_OUT_OF_CAPACITY,
+          })
+        )
+      ),
+      ...documentsClaimedBySchedule.map((doc) => asTaskClaimEvent(doc.id, asOk(doc))),
+      ...documentsRequestedButNotClaimed.map((doc) =>
+        asTaskClaimEvent(
+          doc.id,
+          asErr({
+            task: some(doc),
+            errorType: TaskClaimErrorType.CLAIMED_BY_ID_NOT_IN_CLAIMING_STATUS,
+          })
+        )
+      ),
+      ...documentsRequestedButNotReturned.map((id) =>
+        asTaskClaimEvent(
+          id,
+          asErr({ task: none, errorType: TaskClaimErrorType.CLAIMED_BY_ID_NOT_RETURNED })
+        )
+      ),
+    ]);
+
+    const stats = {
+      tasksUpdated,
+      tasksConflicted,
+      tasksRejected: documentsClaimedByIdOutOfCapacity.length,
+      tasksClaimed: documentsClaimedByIdWithinCapacity.length + documentsClaimedBySchedule.length,
+    };
+
+    if (docs.length !== stats.tasksClaimed + stats.tasksRejected) {
+      this.logger.warn(
+        `[Task Ownership error]: ${stats.tasksClaimed} tasks were claimed by Kibana, but ${
+          docs.length
+        } task(s) were fetched (${docs.map((doc) => doc.id).join(', ')})`
+      );
+    }
+
+    return {
+      stats,
+      docs: [...documentsClaimedByIdWithinCapacity, ...documentsClaimedBySchedule],
+    };
+  };
+
+  private async markAvailableTasksAsClaimed({
```

Comment:
This was extracted from TaskStore and moved here in order to keep the Store focused on ES APIs and keep domain specific business logic separated.